google/cloud/storage/blob.py

```diff
@@ -2006,34 +2006,71 @@ def make_private(self, client=None):
         self.acl.all().revoke_read()
         self.acl.save(client=client)
 
-    def compose(self, sources, client=None, timeout=_DEFAULT_TIMEOUT):
+    def compose(
+        self,
+        sources,
+        client=None,
+        timeout=_DEFAULT_TIMEOUT,
+        if_generation_match=None,
+        if_metageneration_match=None,
+    ):
         """Concatenate source blobs into this one.
 
         If :attr:`user_project` is set on the bucket, bills the API request
         to that project.
 
         :type sources: list of :class:`Blob`
-        :param sources: blobs whose contents will be composed into this blob.
+        :param sources: Blobs whose contents will be composed into this blob.
 
         :type client: :class:`~google.cloud.storage.client.Client` or
                       ``NoneType``
-        :param client: (Optional) The client to use.  If not passed, falls back
+        :param client: (Optional) The client to use. If not passed, falls back
                        to the ``client`` stored on the blob's bucket.
+
         :type timeout: float or tuple
         :param timeout: (Optional) The amount of time, in seconds, to wait
             for the server response.
 
             Can also be passed as a tuple (connect_timeout, read_timeout).
             See :meth:`requests.Session.request` documentation for details.
+
+        :type if_generation_match: list of long
+        :param if_generation_match: (Optional) Make the operation conditional on whether
+                                    the blob's current generation matches the given value.
+                                    Setting to 0 makes the operation succeed only if there
+                                    are no live versions of the blob. Generation numbers list
+                                    should have the same order that ``sources`` have.
+
+        :type if_metageneration_match: list of long
+        :param if_metageneration_match: (Optional) Make the operation conditional on whether the
+                                        blob's current metageneration matches the given value.
+                                        Metageneration numbers list should have the same order
+                                        that ``sources`` have.
         """
         client = self._require_client(client)
         query_params = {}
 
         if self.user_project is not None:
             query_params["userProject"] = self.user_project
 
+        source_objects = []
+        for index, source in enumerate(sources):
+            source_object = {"name": source.name}
+
+            preconditions = {}
+            if if_generation_match is not None:
+                preconditions["ifGenerationMatch"] = if_generation_match[index]
```

Comment:
This pattern will error out with IndexError and a stacktrace internal to our code if the user supplies a list that is too short. Perhaps we should catch the IndexError or check the length in advance, and error out with a ValueError and a useful error message?

Reply:
Yeah, I had thoughts on that, not sure about the best solution. Raising an error in case of different lengths makes sense, but what if user will try to compose 10 blobs? For example, he wanted to use concrete generation for one blob only - in this case he'll have to write `None` for nine others. I've tried to deal with this problem using `iter()` and `next()` in [another related PR](https://github.com/googleapis/python-storage/pull/130#discussion_r420131309), PTAL. Maybe we should do something similar here? (I was going to, actually, but we got on pause)

@frankyn, what do you think?

Reply:
I would provide the following example:

# Option 1
```python

storage_client = storage.Client()

bucket = storage_client.bucket('anima-frank')

list_of_blobs = [bucket.blob('name'), bucket.bloc('name)]
if_generation_match_list = [None] * len(list_of_blobs)
if_generation_match_list[0] = `123`

composed_blob = bucket.blob('composed-name)
composed_blob.compse(list_of_blobs, if_generation_match)
```

It still feels clunky, what about instead doing something like:

# Option 2
```python
list_of_blobs = [{'blob': bucket.blob('name'), bucket.bloc('name), 'if_generation_match': 123), bucket.blob('name2')]


composed_blob = bucket.blob('composed-name)
composed_blob.compse(list_of_blobs)
```

If the element is a dict() then get values?

Reply:
Oh, I thought mix-typed args are usually off the table. But I definitely like the idea

I've done everything locally, but I suppose I'll hold pushing until we merge #123, as I'm using `_add_generation_match_parameters()` in new version, which will be added with pt1. I'd like not to merge #123 into this PR - it'll be +1.5 thousands lines.

![image](https://user-images.githubusercontent.com/26030071/81948547-87c69800-960a-11ea-9f63-cc42421d22c8.png)

Reply:
Thanks @IlyaFaer, let's hold on this PR until early next week to get @crwilcox's feedback. We want to be careful with this surface because it's unlike others in the library. 

Thank you for your patience on this.

Reply:
Hi Ilya,

On discussion we've decided to go ahead with option 1. I've cleared up some ambiguities; see example of UI:

```storage_client = storage.Client()

bucket = storage_client.bucket('anima-frank')

list_of_blobs = [bucket.blob('name1'), bucket.blob('name2')]
if_generation_match_list = [None] * len(list_of_blobs)
if_generation_match_list[0] = `123`

composed_blob = bucket.blob('composed-name)
composed_blob.compse(list_of_blobs, if_generation_match)```

Reply:
Got it. Added an example and error raising. Also I've added one more condition to avoid sending generation match parameters with `None` values (added unit test to check this case)